### ISSUE TITLE
[POC]: Engine Abstraction

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#9602]
Based on the proposal #9601 I will do a POC to verify the idea and maybe uncover some edge cases. I will use this issue to document my findings and certainties and uncertainties. 

Part of #9600

**Comment by Zelldon** [comment]
## Friday - 2022-06-24

The first day of the POC just creating some issues and starting to look at the code.

Notes:

 * Created a branch here https://github.com/camunda/zeebe/tree/zell-poc-engine-abstraction
 * We need a new class for the engine which encapsulates general logic
 * We have to change how current writing is done. This is a huge effort, if we introduce new interface or change the general structure I think. Alternative Idea: it might be easier if we rename/refactor the current writer interfaces, e.g. rename to producers / or record list or something, which can be used to build a list of records, and append new records. This list is then consumed by the StreamProcessor and written to the dispatcher.
 * TypedRecordProcessor has a position in the interface, this can completely removed (done in the branch)
 * TypedRecordProcessor has some todo's to remove the writers -> It seems the processors get the writers in the CTOR's. Might be possible to completely remove it, and the engine creates their own reusable RecordList structure which it returns after processing everytime. StateWriter, could be a listener of the RecordList, so it can react on such new records. This probably needs some more in-depth thinking...

**Comment by Zelldon** [comment]
## Wednesday - 2022-06-29

Again only a half day to work on this topic. What I plan to investigate today:

1. Introduce a different interface RecordListBuilder which is implemented by the writers and used in the engine (so no writer anymore here)
2. Error handling in the engine, callback in the engine if an error appears.

As everyday I will update this comment during my progress.

Notes:

**Comment by Zelldon** [comment]
## Thursday - 2022-06-30

Again half-day of POC. Plan for today, add error handling to the engine, and introduce the SchedulingService abstraction.

Notes: 

* Add a new method to the engine, which is called `onProcessingError` returns also an ProcessingResult (could also be renamed) but this allows to take the same set of actions as after the processing. Means write processing result, commit transaction, send response and start new. 
* :thought_balloon: :question: Thought about the engine interface and the name. I guess we have to find a good name for the engine interface, the engine is more the implementation, but the interface is maybe called StreamProcessor ? :thinking: and the current StreamProcessor could be called StreamingPlatform and we build a distributed streaming platform with Zeebe and partitions :shrug:

**Comment by Zelldon** [comment]
# Summary

## TL;DR;

* It is possible. 
* It is a lot to do, but we can do it incrementally.
* I personally like it, but I might be biased. 
* In general I think it will help a lot to split this into more modules and make a clear boundary
* It could help us in doing some changes in the ActorScheduler
* It will improve our testing
* It allows us to run the same StreamPlatform with a different StreamProcessor, maybe we can reuse it for the Exporting later.

## Details

In the last week, I spent some time doing the POC and implementing Proposal #9601. I was able to implement most of it, of course, some of this is still a bit hacky, and it is not fully working nor complete. It gives a good idea of how it can look in the end, especially how it would improve our boundaries. Furthermore, I see a good possibility to completely split up the StreamPlatform from the engine into separate modules, which allows us also to better each component. 


### How it can look like

_The following order is not necessarily the order how to implement it._

#### Overview:

1. [We rename the now known StreamProcessor into for example StreamPlatform ](https://github.com/camunda/zeebe/commit/d377d081e7148661ecde48a41574ffc288290824)(discussable of course #namingishard) 
2. We introduce a new interface called StreamProcessor, where the Implementation is the engine https://github.com/camunda/zeebe/commit/ce20359349836babee955c28a531bf80185f4d58 https://github.com/camunda/zeebe/commit/8a389e4bc349eaa4edcac7bfd8b02d6bbdd9f69d
3. We could create the engine completely outside of the StreamPlatform and then in the Broker, which makes it possible to decouple  both more
4. This also allows us to test both separately in more detail, with a simple StreamProcessor test implementation

##### Details

###### Writers

One of the bigger parts is removing/replacing the writers in the engine and making it return something. During the POC I came to the conclusion that we can just rename the current writers, since they already interfaces we can reuse. 

* [Renaming the writers to something like builders](https://github.com/camunda/zeebe/commit/47cb438f8b1c6f39d514978b826918a6da8b6a31) (#namingishard lets discuss this) and https://github.com/camunda/zeebe/commit/d91db4508cd382301c218f27aeb5530715120ace
* [Split up the writing to an internal Buffer and writing into the dispatcher into separate classes.](https://github.com/camunda/zeebe/commit/67ecfb29b235e65bf149e1ec74f2181fbda78961)
  * This allows us to create the Builders (BufferWriters) on engine creation (initialization) so they can be always available without issues.
  * The dispatcher writer accepts a buffer writer to write into his buffer and then to the dispatcher (can be of course improved)
  * This means we can create the second writer later and this is only used in the StreamPlatform
* Splitting this up allows us to return the BufferWriter as result of the Engine 
* [We can remove the proxy writer, this is no longer necessary](https://github.com/camunda/zeebe/commit/88ed05a7553409498c15b65aa87e8c5c49f36aff)
* [We can remove the writers from the Processing interface (after the side effects are gone)](https://github.com/camunda/zeebe/commit/154b911c5a0dbf592fcedc915847f714f5474312)
* We have to do the same with the ResponseWriter, but haven't done this in the POC

###### Processing

 * [One interesting details, we need to mark processing more explicit.](https://github.com/camunda/zeebe/commit/4aab0002fc10400b76928af1d912cad110c5c76e)
 * [We can remove the side effects, especially if we have the new SchedulingService](https://github.com/camunda/zeebe/commit/9ccb0f899d49dda5b0ad162263275af105a680a9) and https://github.com/camunda/zeebe/commit/e6773ebfd5faeb82d45b365c58387676865886a9
 * [Move related processing to the engine](https://github.com/camunda/zeebe/commit/12c0595ac6b08637a7254fb2dc3af98b4f86cc6c)
 * [Writing of records are done with new ProcessingResult](https://github.com/camunda/zeebe/commit/67ecfb29b235e65bf149e1ec74f2181fbda78961#diff-f7f005cc99c7e4b3bde9aed2a12c3d77a49f3eea8ab217be4a37586382cd8b01R354)

###### Error Handling

* [Error handling moving into the engine is rather simple, since we have already built it like that in the StreamPlatform before. ](https://github.com/camunda/zeebe/commit/7deca560648eec351f46a8391e05427bcd0d4e64)
* If we also return here a Result, we can use the same path as we do in the processing which is quite elegant.

###### SchedulingService

* [In order to abstract the actor away, we introduce a new abstraction around that, which can be used inside the Engine.](https://github.com/camunda/zeebe/commit/176cfabe3b0997a60604ddb315a9c6607c7b8a40)
* This allows replacing the usage of the ActorControl in other Classes like [DueDateChecker](https://github.com/camunda/zeebe/commit/cb92d3f16820f81000347366beb7fa0f07785c3f)
* [We introduce a support for futures, which is necessary for DeploymentDistributor ](https://github.com/camunda/zeebe/commit/eb9d50e32010d797a0e2fa042f0ef241a42a32b1)
* [We can with that abstraction change remove several code, where we have to schedule, cancel and reschedule timers](https://github.com/camunda/zeebe/commit/fdf6fe453b22999a3522d6a4310f7c10b582bfe4)
* In general it allows us to remove the pausing and resuming from the Lifecycle listeners.

**Comment by npepinpe** [comment]
- Should the engine implementation actually serialize the outputs?
  - Pros: frictionless, no need to interact with the distributed, easier to extend later if we add different processors.
  - Cons: more complicated for testing, embedding, possibly need to implement own integrity checks, schema evolution, not reusing the expertise of the distributed team regarding persistence.

- What if an error occurs in onProcessingError? Do we have some escape hatch? How about UnrecoverableError classes? How much control do we give to the engine (i.e. can it shut itself down on unrecoverable error)?

- Abstracting actor scheduler sounds good. What is the minimal interface? Can we get rid of runUntilDone and the likes, and just go for two capabilities: defer execution (e.g. `run` or `submit`) and `schedule` (e.g. `runDelayed`)? That would greatly simplify things. The caller can take care of `runUntilDone` by implementing their own state machine which ensures no other tasks can be completed in between. However if we want to keep the queuing mechanism, then this is not possible. Do we need it, though?

- Side effects: how to guarantee that a side effect (which may be submitted in between processing) does not affect the state/buffers? It should run in an isolated sandbox in that case.

- Should the stream processor control the flow of processing? e.g. using a credits based approach, it would notify the underlying platform that it can only accept so many credits. So instead of pausing, you could just set the credits to 0. You could also always just have 1 credit which is added after processing, and consumed when processing starts. Might be overengineered for now since we only always have 1 credit.

- Exporters? What happens to those? Are they also stream processors, or do they become a special case? After all they modify the state, which means they're somehow touching the engine (though we can somehow abstract the state to be only "part" of RocksDB and hope they are sufficiently isolated).

**Comment by saig0** [comment]
Minor suggestions regarding the **naming**:
* in the `StreamProcessor` interface: we could rename [apply()](https://github.com/camunda/zeebe/blob/zell-poc-engine-abstraction/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java#L14) to `replay()`
  * use the correct terms: replay (event) + process (command)
* we could rename `*Builders` like [CommandsBuilder](https://github.com/camunda/zeebe/blob/zell-poc-engine-abstraction/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/CommandsBuilder.java) to `*Buffer` like `CommandBuffer`
  * it is is a buffer of commands/events/rejections/*
  * you can append items to a buffer
  * a builder doesn't store anything but return the created object
  * in our case, we use it like a buffer

**Comment by Zelldon** [comment]
Thanks @saig0 and @npepinpe for your feedback and input I moved you comments to the related sub-topics. Feel free to discuss there further with the team. 

The following I think are currently out of scope.


> - Should the stream processor control the flow of processing? e.g. using a credits based approach, it would notify the underlying platform that it can only accept so many credits. So instead of pausing, you could just set the credits to 0. You could also always just have 1 credit which is added after processing, and consumed when processing starts. Might be overengineered for now since we only always have 1 credit.

Not sure whether I get the idea, and yes sounds a bit overengineered right now. 

> - Exporters? What happens to those? Are they also stream processors, or do they become a special case? After all they modify the state, which means they're somehow touching the engine (though we can somehow abstract the state to be only "part" of RocksDB and hope they are sufficiently isolated).

As mentioned earlier we will focus first on the StreamProcessing and Engine part and abstraction. We might be able to run the StreamProcessor later together with both the engine and exporter but this is not clear yet see https://github.com/camunda/zeebe/issues/9725

>
> - I think the proposal from Deepthi regarding rewriting the records is a good idea and will help here, where then only RecordMetadata and RecordData (or whatever) are handled here, without any envelope. This will impact the serialization point, so when starting we should figure out which to do first.

Let's take a look at this after we did most of the needed changes. 


I will close with this the POC, we have a break down of the topic here #9600
